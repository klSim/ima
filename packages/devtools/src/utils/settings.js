import defaultSettings from './defaultSettings';

/**
 * Saves object to chrome.storage.local. Existing settings with the same key
 * are overwritten.
 *
 * @param {object} settings key-value pair object.
 */
function setSettings(settings) {
  chrome.storage.local.set(settings);
}

/**
 * Promisified call to chrome.storage.local.get for getting current settings.
 * Returned settings is merged with default settings json object.
 *
 * @param {?string} key Optional settings key, if null whole storage content is returned.
 * @returns {Promise<object>}
 */
function getSettings(key = null) {
  return new Promise(resolve => {
    chrome.storage.local.get(key, result => {
      if (key) {
        resolve(result[key] !== undefined ? result[key] : defaultSettings[key]);
      } else {
        const { presets: defaultPresets, ...restSettings } = defaultSettings;
        const { presets: resultPresets, ...resultRest } = result;

        // Include default settings
        const settings = {
          ...restSettings,
          ...resultRest,
          presets: {
            ...resultPresets,
            ...defaultPresets,
          },
        };

        resolve(settings);
      }
    });
  });
}

export { setSettings, getSettings };
