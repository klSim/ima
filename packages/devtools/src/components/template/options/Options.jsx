/* eslint-disable react/jsx-no-target-blank */
import uid from 'easy-uid';
import PropTypes from 'prop-types';

import { Button } from '@/components/atom';
import { HookEntryContainer } from '@/components/molecule';
import { PresetsModalContainer } from '@/components/organism';
import { getSettings, setSettings } from '@/utils';

import styles from './options.less';

export default class Options extends React.PureComponent {
  static propTypes() {
    return {
      hookIds: PropTypes.arrayOf(PropTypes.number),
      selectedPresetId: PropTypes.string,
      presets: PropTypes.object,
      setPresets: PropTypes.func,
      addHook: PropTypes.func,
      alertSuccess: PropTypes.func,
    };
  }

  constructor(props) {
    super(props);

    this.state = {
      modalOpened: false,
    };
  }

  get isEditable() {
    const { presets, selectedPresetId } = this.props;

    return !!(
      presets &&
      selectedPresetId &&
      presets[selectedPresetId].editable
    );
  }

  async componentDidMount() {
    const { setPresets } = this.props;
    const { presets, selectedPresetId } = await getSettings();

    setPresets({
      presets,
      selectedPresetId,
    });
  }

  render() {
    const { modalOpened } = this.state;
    const { hookIds } = this.props;

    return (
      <div className={styles.container}>
        <PresetsModalContainer
          onClose={e => this.onModalClose(e)}
          opened={modalOpened}
        />

        <form onSubmit={e => this.onSubmit(e)} className={styles.form}>
          {this._renderHeader()}
          <div className={styles.hooksWrapper}>
            {hookIds &&
              hookIds.map(id => {
                return (
                  <HookEntryContainer
                    editable={this.isEditable}
                    id={id}
                    key={id}
                  />
                );
              })}
          </div>
          {this._renderFooter()}
        </form>
      </div>
    );
  }

  _renderHeader() {
    const { presets, selectedPresetId } = this.props;

    return (
      <div className={styles.header}>
        <div>
          <h1>Hooks settings</h1>
          <p>
            IMA.js developer tools works by wrapping certain methods in a
            proxy-like functions which sends information to the devtools prior
            to it&#x27;s execution.
          </p>
          <p>
            Since wrapping all functions in your application would produce too
            much clutter in the output, we provide default set of hooks which
            you can customize, turn on/off or completely remove to suit your
            needs.{' '}
            <a
              href='https://imajs.io/docs/devtools-introduction'
              target='_blank'
            >
              For more information about devtools, visit imajs.io
            </a>
            .
          </p>
        </div>

        <div className={styles.actionsWrapper}>
          {this.isEditable && (
            <Button onClick={e => this.onAdd(e)} color='primary'>
              Add Hook
            </Button>
          )}
          <div className={styles.loadPresetsWrapper}>
            <p>
              <span className={styles.presetsLabel}>Current preset:</span>{' '}
              {selectedPresetId && presets[selectedPresetId].name}
            </p>
            <Button onClick={e => this.onLoadPreset(e)}>
              Load or Create Preset
            </Button>
          </div>
        </div>
      </div>
    );
  }

  _renderFooter() {
    return (
      <div className={styles.footer}>
        {!this.isEditable && (
          <p className={styles.notEditable}>
            Current preset is not editable, please create new one by clicking on
            &#x27;Load or Create Preset&#x27; button.
          </p>
        )}
        <Button className={styles.saveBtn} color='success' type='submit'>
          Save Changes
        </Button>
      </div>
    );
  }

  onAdd(e) {
    e.preventDefault();
    const { addHook } = this.props;

    addHook(this._createHook());
  }

  onLoadPreset(e) {
    e.preventDefault();

    this.setState({
      modalOpened: true,
    });
  }

  onModalClose() {
    this.setState({
      modalOpened: false,
    });
  }

  onSubmit(e) {
    e.preventDefault();

    const { presets, selectedPresetId, alertSuccess } = this.props;
    const groupedData = {};
    const formData = new FormData(e.target);

    // Get changes from state
    for (let [key, value] of formData.entries()) {
      const [name, id] = key.split('__');

      if (!groupedData[id]) {
        groupedData[id] = {};
      }

      groupedData[id][name] = value;
    }

    const hooks = presets[selectedPresetId].hooks;

    // Merge with global redux state
    for (let id in hooks) {
      groupedData[id] = {
        ...hooks[id],
        ...groupedData[id],
        opened: false,
      };
    }

    // Create new presets object with new data
    const newPresets = { ...presets };
    newPresets[selectedPresetId] = {
      ...newPresets[selectedPresetId],
      ...{ hooks: groupedData },
    };

    // Save settings to storage
    setSettings({
      presets: newPresets,
      selectedPresetId,
    });

    alertSuccess('Changes were saved.');
  }

  _createHook() {
    const id = uid();
    const displayId = id.substring(0, 6);

    return {
      id,
      enabled: false,
      opened: false,
      name: `Hook - ${displayId}`,
      description: `Description for hook - ${displayId}`,
      code: '',
    };
  }
}
