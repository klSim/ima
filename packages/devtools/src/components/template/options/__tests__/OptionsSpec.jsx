import React from 'react';
import { shallow } from 'enzyme';
import Options from '../Options';
import * as settings from 'services/settings';

jest.mock('easy-uid');
import uid from 'easy-uid';

describe('Options template', () => {
  let wrapper, instance;

  const event = {
    preventDefault: jest.fn(),
  };

  const props = {
    presets: {
      0: {
        id: '0',
        name: 'name',
        editable: true,
        selected: false,
      },
    },
    setPresets: jest.fn(),
    addHook: jest.fn(),
    alertSuccess: jest.fn(),
    selectedPresetId: '0',
    hookIds: ['1', '2', '3'],
  };

  beforeAll(() => {
    global.chrome = {
      storage: {
        local: {
          get: jest.fn,
        },
      },
    };
  });

  beforeEach(() => {
    wrapper = shallow(<Options {...props} />);
    instance = wrapper.instance();
    event.preventDefault.mockClear();
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot with modal opened', () => {
    wrapper.setState({
      modalOpened: true,
    });

    expect(wrapper).toMatchSnapshot();
  });

  describe('get isEditable', () => {
    it('should return true if preset is editable', () => {
      expect(instance.isEditable).toBe(true);
    });

    it('should return false if preset is not editable', () => {
      wrapper.setProps({
        presets: {
          ...props.presets,
          0: {
            ...props.presets['0'],
            editable: false,
          },
        },
      });

      expect(instance.isEditable).toBe(false);
    });

    it("should return false if selected preset doesn't exist", () => {
      wrapper.setProps({
        selectedPresetId: null,
      });

      expect(instance.isEditable).toBe(false);
    });
  });

  describe('componentDidMount', () => {
    it('should fetch and set settings on mount', async () => {
      jest
        .spyOn(settings, 'getSettings')
        .mockImplementation()
        .mockImplementation(() => ({
          presets: 'settingsPresets',
          selectedPresetId: '0',
        }));

      await instance.componentDidMount();

      expect(settings.getSettings.mock.calls).toHaveLength(1);
      expect(instance.props.setPresets.mock.calls).toHaveLength(1);
      expect(instance.props.setPresets.mock.calls[0][0]).toStrictEqual({
        presets: 'settingsPresets',
        selectedPresetId: '0',
      });
    });
  });

  describe('onAdd', () => {
    it('should call props.addHook with generated hook', () => {
      jest
        .spyOn(instance, '_createHook')
        .mockImplementation()
        .mockReturnValue('newHook');

      instance.onAdd(event);

      expect(event.preventDefault.mock.calls).toHaveLength(1);
      expect(instance.props.addHook.mock.calls).toHaveLength(1);
      expect(instance.props.addHook.mock.calls[0][0]).toBe('newHook');
      expect(instance._createHook.mock.calls).toHaveLength(1);
    });
  });

  describe('onLoadPreset', () => {
    it('should open presets modal window', () => {
      jest.spyOn(instance, 'setState').mockImplementation();

      instance.onLoadPreset(event);

      expect(event.preventDefault.mock.calls).toHaveLength(1);
      expect(instance.setState.mock.calls).toHaveLength(1);
      expect(instance.setState.mock.calls[0][0]).toStrictEqual({
        modalOpened: true,
      });
    });
  });

  describe('onModalClose', () => {
    it('should closeModalWindow', () => {
      jest.spyOn(instance, 'setState').mockImplementation();

      instance.onModalClose();

      expect(instance.setState.mock.calls).toHaveLength(1);
      expect(instance.setState.mock.calls[0][0]).toStrictEqual({
        modalOpened: false,
      });
    });
  });

  describe('onSubmit', () => {
    beforeEach(() => {
      jest.spyOn(settings, 'setSettings').mockImplementation();
      global.FormData = function () {
        return {
          entries: jest.fn().mockImplementation(() => {
            return [['name__0', 'newName']];
          }),
        };
      };

      instance.props.alertSuccess.mockClear();
    });

    it('should call set settings with new extracted data', () => {
      instance.onSubmit(event);

      expect(settings.setSettings.mock.calls).toHaveLength(1);
      expect(settings.setSettings.mock.calls[0][0]).toStrictEqual({
        presets: {
          ...props.presets,
          0: {
            ...props.presets['0'],
            hooks: {
              0: {
                name: 'newName',
              },
            },
          },
        },
        selectedPresetId: props.selectedPresetId,
      });
    });

    it('should show success alert', () => {
      instance.onSubmit(event);

      expect(event.preventDefault.mock.calls).toHaveLength(1);
      expect(instance.props.alertSuccess.mock.calls).toHaveLength(1);
      expect(instance.props.alertSuccess.mock.calls[0][0]).toBe(
        'Changes were saved.'
      );
    });
  });

  describe('_createHook', () => {
    it('should create blank hook object with generated ID', () => {
      uid.mockReturnValue('2fghzj-123456');

      expect(instance._createHook()).toStrictEqual({
        id: '2fghzj-123456',
        enabled: false,
        opened: false,
        name: 'Hook - 2fghzj',
        description: 'Description for hook - 2fghzj',
        code: '',
      });
    });
  });
});
