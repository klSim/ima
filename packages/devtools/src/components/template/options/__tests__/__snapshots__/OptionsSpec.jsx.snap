// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options template should match snapshot 1`] = `
<div
  className="container"
>
  <Connect(PresetsModal)
    onClose={[Function]}
    opened={false}
  />
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="header"
    >
      <div>
        <h1>
          Hooks settings
        </h1>
        <p>
          IMA.js developer tools works by wrapping certain methods in a proxy-like functions which sends information to the devtools prior to it's execution.
        </p>
        <p>
          Since wrapping all functions in your application would produce too much clutter in the output, we provide default set of hooks which you can customize, turn on/off or completely remove to suit your needs.
           
          <a
            href="https://imajs.io/docs/devtools-introduction"
            target="_blank"
          >
            For more information about devtools, visit imajs.io
          </a>
          .
        </p>
      </div>
      <div
        className="actionsWrapper"
      >
        <Button
          className=""
          color="primary"
          onClick={[Function]}
        >
          Add Hook
        </Button>
        <div
          className="loadPresetsWrapper"
        >
          <p>
            <span
              className="presetsLabel"
            >
              Current preset:
            </span>
             
            name
          </p>
          <Button
            className=""
            onClick={[Function]}
          >
            Load or Create Preset
          </Button>
        </div>
      </div>
    </div>
    <div
      className="hooksWrapper"
    >
      <Connect(HookEntry)
        editable={true}
        id="1"
        key="1"
      />
      <Connect(HookEntry)
        editable={true}
        id="2"
        key="2"
      />
      <Connect(HookEntry)
        editable={true}
        id="3"
        key="3"
      />
    </div>
    <div
      className="footer"
    >
      <Button
        className="saveBtn"
        color="success"
        type="submit"
      >
        Save Changes
      </Button>
    </div>
  </form>
</div>
`;

exports[`Options template should match snapshot with modal opened 1`] = `
<div
  className="container"
>
  <Connect(PresetsModal)
    onClose={[Function]}
    opened={true}
  />
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="header"
    >
      <div>
        <h1>
          Hooks settings
        </h1>
        <p>
          IMA.js developer tools works by wrapping certain methods in a proxy-like functions which sends information to the devtools prior to it's execution.
        </p>
        <p>
          Since wrapping all functions in your application would produce too much clutter in the output, we provide default set of hooks which you can customize, turn on/off or completely remove to suit your needs.
           
          <a
            href="https://imajs.io/docs/devtools-introduction"
            target="_blank"
          >
            For more information about devtools, visit imajs.io
          </a>
          .
        </p>
      </div>
      <div
        className="actionsWrapper"
      >
        <Button
          className=""
          color="primary"
          onClick={[Function]}
        >
          Add Hook
        </Button>
        <div
          className="loadPresetsWrapper"
        >
          <p>
            <span
              className="presetsLabel"
            >
              Current preset:
            </span>
             
            name
          </p>
          <Button
            className=""
            onClick={[Function]}
          >
            Load or Create Preset
          </Button>
        </div>
      </div>
    </div>
    <div
      className="hooksWrapper"
    >
      <Connect(HookEntry)
        editable={true}
        id="1"
        key="1"
      />
      <Connect(HookEntry)
        editable={true}
        id="2"
        key="2"
      />
      <Connect(HookEntry)
        editable={true}
        id="3"
        key="3"
      />
    </div>
    <div
      className="footer"
    >
      <Button
        className="saveBtn"
        color="success"
        type="submit"
      >
        Save Changes
      </Button>
    </div>
  </form>
</div>
`;
